### PR TITLE
Character List

### PR DESCRIPTION
**Pull request by JevLOMCN**
Added a Character List which will display all Characters.

Before this the process of searching Characters was to either go through them individually using the Account Info Form or the Players Online Tab.

Also has a Filter By Item UID box which allows you to quickly identify the owner of an Item.

### CODE REVIEW

File: Server/MirEnvir/Envir.cs [hunk 1]
```diff
@@ -2532,7 +2533,93 @@ public bool CharacterExists(string name)
...
+            }
+
+            return list;
+        }
+        public List<CharacterInfo> MatchPlayerbyItem(string ItemID, bool match = false)
+        {
+            if (string.IsNullOrEmpty(ItemID)) return new List<CharacterInfo>(CharacterList);
+
+            ulong itemId = 0;
+            ulong.TryParse(ItemID, out itemId);
```
Comment (Suprcode) [inline Server/MirEnvir/Envir.cs:2563]: you should probably just return nothing if the id isn't a number

File: Server/MirEnvir/Envir.cs [hunk 2]
```diff
@@ -2532,7 +2533,93 @@ public bool CharacterExists(string name)
...
+
+            List<CharacterInfo> list = new List<CharacterInfo>();
+
+            for (int i = 0; i < CharacterList.Count; i++)
```
Comment (Suprcode) [inline Server/MirEnvir/Envir.cs:2567]: i dont get the difference between match and not a match?

File: Server.MirForms/SMain.cs [hunk 3]
```diff
@@ -558,5 +559,14 @@ private void heroesToolStripMenuItem_Click(object sender, EventArgs e)
...
             form.ShowDialog();
         }
+
+        private void CharacterToolStripMenuItem_Click(object sender, EventArgs e)
+        {
+            CharacterInfoForm form = new CharacterInfoForm();
+
+            form.ShowDialog();
+
+            form.Dispose();
```
Comment (Suprcode) [inline Server.MirForms/SMain.cs:569]: why are you disposing it straight after showing it?

File: Server/MirEnvir/Envir.cs [hunk 4]
```diff
@@ -2532,7 +2533,93 @@ public bool CharacterExists(string name)
...
+                    foreach (var item in CharacterList[i].AccountInfo.Storage)
+                        if (item != null && item.UniqueID == itemId && !list.Contains(CharacterList[i]))
+                            list.Add(CharacterList[i]);
+                }
+
+            }
+
+            return list;
+        }
+        public AccountInfo GetAccountArchived(string accountID)
```
Comment (Suprcode) [inline Server/MirEnvir/Envir.cs:2615]: method isn't called by anything?

**Review by Suprcode — CHANGES_REQUESTED** [review]
Several incomplete features and weird code issues to fix here. No need to rush pull requests through until it 100% works as intended.

Also please provide screenshots when forms are changed/added so its easier to review

File: Server.MirForms/SMain.cs (hunk 3 above)
Comment (JevLOMCN) [reply Server.MirForms/SMain.cs:569]: Fixed

File: Server/MirEnvir/Envir.cs (hunk 4 above)
Comment (JevLOMCN) [reply Server/MirEnvir/Envir.cs:2615]: Removed

File: Server/MirEnvir/Envir.cs (hunk 1 above)
Comment (JevLOMCN) [reply Server/MirEnvir/Envir.cs:2563]: I've just removed the null check. If a letter is input then it returns nothing and if a unused item id is input it does the same.

File: Server/MirEnvir/Envir.cs (hunk 2 above)
Comment (JevLOMCN) [reply Server/MirEnvir/Envir.cs:2567]: When match is set to false, the method does not check:

AccountInfo.Storage: Excludes checking items stored in the account storage.
Mail: Excludes checking items attached to mails.

File: Server/MirEnvir/Envir.cs (hunk 1 above)
Comment (Suprcode) [reply Server/MirEnvir/Envir.cs:2561]: yes but why bother doing the rest of the code when you know 0 will never find anything?

if its not parsed as a number, or if 0 is entered - it should just return no results straight away

Comment (JevLOMCN) [reply Server/MirEnvir/Envir.cs:2561]: I see, I'll do a pr

File: Server/MirEnvir/Envir.cs (hunk 2 above)
Comment (Suprcode) [reply Server/MirEnvir/Envir.cs:2565]: In that case just wrap the check around those loops which the "match" utilises, so you don't have to duplicate the entire code

File: Server/MirEnvir/Envir.cs (hunk 4 above)
Comment (Suprcode) [reply Server/MirEnvir/Envir.cs]: The archived list at the top also needs removing please as that seems to have no use

File: Server/MirEnvir/Envir.cs (hunk 1 above)
Comment (JevLOMCN) [reply Server/MirEnvir/Envir.cs:2561]: image

File: Server/MirEnvir/Envir.cs (hunk 4 above)
Comment (JevLOMCN) [reply Server/MirEnvir/Envir.cs]: Fixed

File: Server/MirEnvir/Envir.cs (hunk 2 above)
Comment (JevLOMCN) [reply Server/MirEnvir/Envir.cs:2565]: I have no clue how

File: Server/MirEnvir/Envir.cs (hunk 1 above)
Comment (Suprcode) [reply Server/MirEnvir/Envir.cs:2561]: Move the list above it and just return the empty list

File: Server/MirEnvir/Envir.cs (hunk 2 above)
Comment (Suprcode) [reply Server/MirEnvir/Envir.cs:2566]: get rid of the if(match) else and and wrap the foreach loops which you only want to be used when if(match) is true

Comment (IHaveALongName) [reply Server/MirEnvir/Envir.cs:2565]: Fixed

**Comment by IHaveALongName** [comment]
> Several incomplete features and weird code issues to fix here. No need to rush pull requests through until it 100% works as intended.
> 
> Also please provide screenshots when forms are changed/added so its easier to review

* Can now search items by Name & UniqueID

image

**Pull request merged**